src/class-bulk-task.php

```diff
@@ -0,0 +1,238 @@
+	}
+
+	/**
+	 * Actions to take after a bulk task is run.
+	 */
+	protected function after_run(): void {
+		wp_defer_term_counting( false );
+		$this->progress?->set_finished();
+	}
+
+	/**
+	 * Actions to take before a bulk task is run.
+	 */
+	protected function before_run(): void {
+		wp_defer_term_counting( true );
+		$this->progress?->set_total( $this->max_id );
+	}
+
+	/**
+	 * Manipulate the WHERE clause of a bulk task query to paginate by ID.
+	 *
+	 * This checks the object hash to ensure that we don't manipulate any other
+	 * queries that might run during a bulk task.
+	 *
+	 * @param  string   $where The WHERE clause of the query.
+	 * @param  WP_Query $query The WP_Query instance (passed by reference).
+	 *
+	 * @return string WHERE clause with our pagination added.
+	 */
+	public function filter__posts_where( $where, $query ) {
+		if ( spl_object_hash( $query ) === $this->object_hash ) {
+			global $wpdb;
+
+			return sprintf(
+				'AND %s.ID > %d AND %s.ID <= %d %s',
+				$wpdb->posts,
+				$this->min_id,
+				$wpdb->posts,
+				$this->min_id + self::STEPPING,
+				$where
+			);
+		}
+
+		return $where;
+	}
+
+	/**
+	 * Loop through any number of posts efficiently with a callback, and output
+	 * the progress.
+	 *
+	 * @param array    $args {
+	 *     WP_Query args. Some have overridden defaults, and some are fixed.
+	 *     Anything not mentioned below will operate as normal.
+	 *
+	 *     @type bool   $ignore_sticky_posts Always true.
+	 *     @type bool   $no_found_rows       Always true.
+	 *     @type string $order               Always 'ASC'.
+	 *     @type string $orderby             Always 'ID'.
+	 *     @type int    $paged               Always 1.
+	 *     @type string $post_status         Defaults to 'any'.
+	 *     @type string $post_type           Defaults to 'any'.
+	 *     @type int    $posts_per_page      Defaults to 100.
+	 *     @type bool   $suppress_filters    Always false.
+	 * }
+	 * @param callable $callable Callback function to invoke for each post.
+	 *                           The callable will be passed a post object.
+	 */
+	public function run( array $args, callable $callable ): void {
+		global $wpdb;
+
+		// Apply default arguments.
+		$args = wp_parse_args(
+			$args,
+			[
+				'post_type'      => 'any',
+				'post_status'    => 'any',
+				'posts_per_page' => 100,
+			],
+		);
+
+		// Force some arguments and don't let them get overridden.
+		$args['ignore_sticky_posts'] = true;
+		$args['no_found_rows']       = true;
+		$args['order']               = 'ASC';
+		$args['orderby']             = 'ID';
+		$args['paged']               = 1;
+		$args['suppress_filters']    = false;
+
+		// Set the min ID from the cursor.
+		$this->min_id = $this->cursor->get();
+
+		// Set the max ID from the database.
+		$this->max_id = $wpdb->get_var( 'SELECT MAX(ID) FROM ' . $wpdb->posts ); // phpcs:ignore WordPress.DB.DirectDatabaseQuery.DirectQuery, WordPress.DB.DirectDatabaseQuery.NoCaching
```

Comment:
Agreed, I'd love to introduce elements like this in future versions to really optimize and speed things up. Even as written, though, traversing 1,000,000 posts 10,000 at a time where there are no found results is pretty quick (typically a few seconds, in my testing). That's especially true if the auto_increment number is set, leaving a hole in the database, because the ID part of the query runs first, and very quickly gets to a null result and moves to the next chunk in the loop.

Reply:
we gamed that scenario out and decided that that edge-of-the-edge case wasn't worth the added complexity (and as kevin mentioned, the cost for that edge case is just a few seconds of empty querying). the compromise was making STEPPING 100x the posts_per_page instead of equal to it (there's a bug around that, which I'll call out elsewhere)